### PR TITLE
PR230: stabilization of runtime introspection contracts and UI presentation

### PR DESCRIPTION
**Pull request by mpieniak01**
## Summary
- stabilize model introspection runtime contract and timeline semantics across backend and frontend
- restore advanced internals data presentation (token flow/checkpoints/attention/saliency) with clear proxy vs native semantics
- add runtime switch telemetry visibility (`runtime_switch_policy`, `last_runtime_switch`) to active runtime endpoints
- extend test coverage for introspection services/view-model and runtime switch telemetry

## Key changes
- backend: analysis/probe/logit-lens/attention/saliency services and introspection route shaping
- frontend: dashboard, sections, view-model, mechanism control, stream handling and i18n strings
- runtime: multi-runtime service adjustments for introspection payload flow
- tests: backend + frontend unit coverage for new semantics and regressions

## Validation
- `make pr-fast` passed locally
- pre-commit hooks passed on commit

## Risk/notes
- large cross-layer change set (backend + frontend + runtime)
- requires live smoke verification on `gemma` and `qwen` runtime profiles after service restart to confirm native/proxy signal mix

## Notatki do wydania

* **New Features**
  * Dodano tryb saliency do analizy introspekcji modelu z obliczaniem czułości na podstawie gradientów
  * Obsługa surowych tokenów w panelu Logit Lens obok znormalizowanych widoków
  * Automatyczne przybliżenie (proxy) dla uwagi i logitów gdy tryb bezpośredni niedostępny
  * Śledzenie ostatniego zdarzenia przełączenia runtime

* **Documentation**
  * Aktualizacja umowy runtime dla introspekcji modelu uwzględniająca nowe możliwości

* **Chores**
  * Dodano bramkę kontroli jakości pre-push

[Review Change Stack](https://app.coderabbit.ai/change-stack/mpieniak01/Venom/pull/499?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: services/multi_runtime/main.py [hunk 1]
```diff
@@ -1225,6 +1281,58 @@ def _run_probe_sync(
...
+        if logits is None:
+            raise RuntimeError("logits_unavailable")
+        next_token_logits = logits[0, -1, :]
+        target_token_id, target_token = _resolve_saliency_target_token(
+            logits_vector=next_token_logits,
+            tokenizer=tokenizer,
+            target_output_token_index=target_output_token_index,
+        )
+        target_logit = next_token_logits[target_token_id]
+        target_logit.backward()
```
[the remark this answers is not in the document]
Comment (mpieniak01) [reply services/multi_runtime/main.py]: Fixed in 678f6fe5. Added `_PROBE_SALIENCY_LOCK` and wrapped gradient-based saliency block (`zero_grad` + `backward`) with the lock to prevent concurrent gradient interference on shared model instances.

File: services/multi_runtime/main.py [hunk 2]
```diff
@@ -1192,13 +1193,68 @@ def _extract_logits_top(
...
+    top_k = min(vocab_size, max(1, rank_index + 1))
+    _values, indices = torch.topk(logits_vector, k=top_k)
+    token_id = int(indices[rank_index] if rank_index < top_k else indices[0])
+    token = str(token_id)
+    try:
+        decoded = tokenizer.convert_ids_to_tokens([token_id])
+        if isinstance(decoded, list) and decoded:
+            token = str(decoded[0])
+    except Exception:
+        pass
```
[the remark this answers is not in the document]
Comment (mpieniak01) [reply services/multi_runtime/main.py]: Fixed in 678f6fe5. Replaced bare `except Exception` with a bounded exception set (`AttributeError`, `TypeError`, `ValueError`, `KeyError`, `IndexError`) in token decode helpers.

File: venom_core/services/model_introspection_probe_service.py [hunk 3]
```diff
@@ -759,17 +764,21 @@ async def run_model_introspection_probe(
...
     }
-    if target_output_token_index is not None:
-        probe_payload["target_output_token_index"] = int(target_output_token_index)
+    if normalized_mode == "saliency" and isinstance(target_output_token_index, int):
+        probe_payload["target_output_token_index"] = max(0, target_output_token_index)
+    # Keep request-level validation for compatibility with callers that still pass
+    # legacy fields, but do not forward unsupported keys to multi_runtime.
+    _ = sanitized_heads
+    if normalized_mode != "saliency":
+        _ = target_output_token_index
```
[the remark this answers is not in the document]
Comment (mpieniak01) [reply venom_core/services/model_introspection_probe_service.py]: Fixed in 678f6fe5. Removed redundant dummy assignments; request validation remains intact while unsupported keys are simply not forwarded.

File: venom_core/infrastructure/traffic_control/config.py [hunk 4]
```diff
@@ -194,6 +201,17 @@ def should_enter_degraded_state(
...
+    def is_provider_exempt_from_degraded_mode(self, provider: str) -> bool:
+        """Sprawdza, czy provider ma bypass globalnego degraded mode."""
+        normalized = provider.strip().lower()
+        if not normalized:
+            return False
+        return normalized in {
+            entry.strip().lower()
+            for entry in self.degraded_mode_exempt_providers
+            if entry.strip()
+        }
```
[the remark this answers is not in the document]
Comment (mpieniak01) [reply venom_core/infrastructure/traffic_control/config.py]: Fixed in 678f6fe5. Added cached normalized exempt-provider set in `TrafficControlConfig` (`PrivateAttr` + `model_post_init`) and switched lookup to O(1) membership without rebuilding the set per call.
